### PR TITLE
feat(PFD): flashing of STD/QNH at TL/TA

### LINKED ISSUE #3293: QNH/STD not blinking when passing Transition Alt
**Mod Version**
Custom FWB c705061

**Describe the bug**
I've set both the Transition Altitude and Level in the FMC, but when passing them the STD / QNH box doesn't blink / flash on the PFD. In fact I never saw it flashing ever, even with other versions (stable, dev etc).

Yesterday I removed the A32NX folder and reinstalled the mod again from scratch, with the same result.

**To Reproduce**
1. Load flight plan from Simbrief - for example going from LOWW to LOWI
2. Set Transition Altitude at 10000 and Transition Level at 11000 in the PERF TAKE OFF and APPR pages.
3. In OPTIONS > FMGC > I have Default Baro set to *AUTO or *HPA (tried both) and Weight Unit to *kg
4. Set QNH to 1025 before take off
5. Set QNH to STD after passing the TA

**Expected behavior**
QNH 1025 should start flashing in the PFD between 10000 and 11000 when climbing
STD should start flashing in the PFD when descending between 110000 and 10000

**Actual behavior**
QNH 1025 and STD (in the PDF) won't flash/blink

**References**

**Additional context**

Was this working before/when did the issue start occurring?
Never saw it working. I'm using the mod for several months.

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub): elastisteve#7446

### PR DESCRIPTION
**Pull request by lukecologne**
Fixes #3293

## Summary of Changes

* Adds flashing of the baro setting indicator when passing TL/TA

A TL/TA value of 0 is treated as no value set. This is a limitation of the FMC code, as it will not initialize the Lvar properly, so it will stay it it's default value of 0 when not entered.

## Additional context

This is a react adaptation of the code in #4806. Supersedes #4806.

Discord username (if different from GitHub): lukecologne#1156

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by MisterChocker — APPROVED** [review]
LGTM

**Comment by frankkopp** [comment]
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #4892
Tier of Testing: 2
Date: 24.5.2021

Testing Process: (Detail is needed)
- Starting flight from runway
- Importing Simbrief route - setting up MCDU as per SOP (CRZ 11.000)
- TA/TL = 5.000 for Origin and 7.000 for Destination
- Take off and climb - TA=5.000 - then climp 11.000
- Descent to 7.000 - then 5.000

Negatives:
I assume that this is not correct - a RL pilot would need to confirm:
When TA is 5.000 and initial climb is 5.000 QNH should not blink even if alt is slightly above
5.000 after leveling. Currently it does blink when plane is minimally above 5.000.
Imho it should only blink if FCU ALT is >5.000 (TA).
Same is true when descending: TL FL070 - descending to FL70 - QNH blinks when slightly below 
FL70 although FCU still is on FL070
Imho it should only blink if FCU ALT is <7.000 (TL).

Testing Results: Not Passed (if my assumption above is correct)

Conclusions: Blinking otherwise works as expected when clearly above or below TA/TL

Media: Videos available if required

**Comment by frankkopp** [comment]
[6:40 PM] luke [Z+2]: as I said, only the baro alt is relevant
[6:41 PM] luke [Z+2]: at least that is how it's indicated in all the docs
[6:42 PM] Cdr_Maverick [Z+2]: ok - then my assumption was wrong - in consequence it will blink if FCU ALT = TA and plane will level off there as the plane always will be a little above or below 5.000 - then my conclusion is that it works as intended :slight_smile:

Testing Results: Passed 

Conclusions: LGTM

**Comment by HPringles** [comment]
Quality Assurance Trainee Report

Object of testing: #4892 
Tier of Testing: 1
Date: 24/05/2020

Testing Process: 

- Flight Plan Via simbrief entered into MCDU as per SOP (EGKK-EGSS)
- Transition altitudes set as following: Departure: 6000, Approach: 5000
- Initially climbed to 5000, then 6000, then cruise altitude of FL70
- As soon as altitude went above the 6000 TA, QNH starts blinking as expected until set to STD, then stops as expected
- On Approach, as soon as I descend past 5000, STD starts blinking as expected, until set to local QNH, then stops as expected

Negatives: N/A
Testing Results: Passed

**Comment by Samueleonard** [comment]
Quality Assurance Tester Report

Discord: @1samueleonard1
Object of testing: (#4892)
Tier of Testing: (1)
Date: (24/05/2021)

Testing Process: 

loaded in on runway, put basic flight plan in (gcts/gcrr trans alt 5000 (issue)), 
took off, climbed to 5000 (TL), 
levelled at 5000 (no flashing as expected)
changed TA to 4000  on appr page(flashing as expected)
changed TA back to 5000 (flashing, not sure if expected? stops and starts are passing 5000
changed TA to 6000 (expected no flashing however continued to flash even in QNH)
climbed to 15k (cruise)
set TA to 17k (above current, expected flashing but no flashing)
moved TA to 6k and began descent (should flash at 6k)
crossed through TA (below 6k) (flashing as expected) 
changed TA to 4000 (flashing stopped as expected, started at 4000 as expected)
raised TA back to 6000 (flashing continued)
changed to qnh (flashing stopped)

Negatives:  default transition altitudes are not correct (not in scope of this pr though) and issues as described above (need confirmation if they are correct, even though they are edge cases and not normal behaviour)
Testing Results: (Not Passed)

Conclusions: kinda works but needs works still

**Comment by frankkopp** [comment]
Confirmation by pilot that QNH will be flashing at TA even when plane leveled out (FCU==TA).
FlyingCarpet7: "EGLL is a god example. Like the BPK 7F SID .  Initial climb is to 6000 ft and to maintain the alt for quite some time. Transition Altitude is also 6000 ft. There we will get the QNH flashing at us. Sometimes we remember to think ahead and set the trans alt on TO perf page to something like 6050, to avoid the flashing."

File: src/instruments/src/PFD/AltitudeIndicator.jsx [hunk 1]
```diff
@@ -162,10 +162,13 @@ const LinearDeviationIndicator = ({ linearDeviation, alt }) => {
...
 
-const AltimeterIndicator = ({ mode }) => {
+const AltimeterIndicator = ({ mode, alt }) => {
+    const phase = getSimVar('L:A32NX_FMGC_FLIGHT_PHASE', 'enum');
+    const transAlt = getSimVar(phase <= 3 ? 'L:AIRLINER_TRANS_ALT' : 'L:AIRLINER_APPR_TRANS_ALT', 'number');
+
     if (mode === 'STD') {
         return (
-            <g id="STDAltimeterModeGroup">
+            <g id="STDAltimeterModeGroup" className={(alt < transAlt && transAlt !== 0) ? 'BlinkInfinite' : ''}>
```
Comment (Kimbyeoungjang) [inline src/instruments/src/PFD/AltitudeIndicator.jsx:171]:
```suggestion
            <g id="STDAltimeterModeGroup" className={(alt < transAlt && transAlt !== 0 && !Simplane.getIsGrounded()) ? 'BlinkInfinite' : ''}>
```

File: src/instruments/src/PFD/AltitudeIndicator.jsx [hunk 2]
```diff
@@ -182,7 +185,7 @@ const AltimeterIndicator = ({ mode }) => {
     }
 
     return (
-        <g id="AltimeterGroup">
+        <g id="AltimeterGroup" className={(alt > transAlt && transAlt !== 0) ? 'BlinkInfinite' : ''}>
```
Comment (Kimbyeoungjang) [inline src/instruments/src/PFD/AltitudeIndicator.jsx:188]:
```suggestion
        <g id="AltimeterGroup" className={(alt > transAlt && transAlt !== 0 && !Simplane.getIsGrounded()) ? 'BlinkInfinite' : ''}>
```

**Review by Kimbyeoungjang — COMMENTED** [review]
you need to add these code because it should not flashing when on ground

File: src/instruments/src/PFD/AltitudeIndicator.jsx (hunk 1 above)
Comment (Kimbyeoungjang) [inline src/instruments/src/PFD/AltitudeIndicator.jsx:171]:
```suggestion
            <g id="STDAltimeterModeGroup" className={(alt <= transAlt && transAlt !== 0) ? 'BlinkInfinite' : ''}>
```

File: src/instruments/src/PFD/AltitudeIndicator.jsx (hunk 2 above)
Comment (Kimbyeoungjang) [inline src/instruments/src/PFD/AltitudeIndicator.jsx:188]:
```suggestion
        <g id="AltimeterGroup" className={(alt >= transAlt && transAlt !== 0) ? 'BlinkInfinite' : ''}>
```

**Review by Kimbyeoungjang — CHANGES_REQUESTED** [review]
also change these > < to <= >=

**Comment by Kimbyeoungjang** [comment]
> Quality Assurance Tester Report
> 
> Discord: @1samueleonard1
> Object of testing: (#4892)
> Tier of Testing: (1)
> Date: (24/05/2021)
> 
> Testing Process:
> 
> loaded in on runway, put basic flight plan in (gcts/gcrr trans alt 5000 (issue)),
> took off, climbed to 5000 (TL),
> levelled at 5000 (no flashing as expected)
> changed TA to 4000 on appr page(flashing as expected)
> changed TA back to 5000 (flashing, not sure if expected? stops and starts are passing 5000
> changed TA to 6000 (expected no flashing however continued to flash even in QNH)
> climbed to 15k (cruise)
> set TA to 17k (above current, expected flashing but no flashing)
> moved TA to 6k and began descent (should flash at 6k)
> crossed through TA (below 6k) (flashing as expected)
> changed TA to 4000 (flashing stopped as expected, started at 4000 as expected)
> raised TA back to 6000 (flashing continued)
> changed to qnh (flashing stopped)
> 
> Negatives: default transition altitudes are not correct (not in scope of this pr though) and issues as described above (need confirmation if they are correct, even though they are edge cases and not normal behaviour)
> Testing Results: (Not Passed)
> 
> Conclusions: kinda works but needs works still

your test seems wrong because you completely ignore flight phase.

you can't change departure airport TA after takeoff and you change destination airport TA.

and you're QNH bar flashing. this mean your phase pass the descend phase.

so function is not working as expected.

**Comment by Samueleonard** [comment]
@Kimbyeoungjang yes me and luke had a chat on discord about the flight phases, and it explains all of the things i saw, so actually the PR is correct, and matches what i saw. I shouldve edited my comment to reflect that sorry

**Comment by easyjetsimpilot** [comment]
Quality Assurance Tester Report

Discord: EasyjetSimPilot
Object of testing: #4892 
Tier of Testing: (2)
Date: 25/5/21

Testing Process: 

Departed LIPZ with Trans Level set to 3000ft on PERF (for departure) & Trans Alt set to 5000ft (for arrival)
After passing 3000ft, 'QNH' flashes as expected on PFD
Completed a climb up to FL100
Descended down to 3000ft. On passing 5000ft, STD flashes as expected on PFD

Negatives: N/A
Testing Results: Passed

Conclusions: Works as expected. Well done and Thank you

**Comment by DavidEGTE** [comment]
Quality Assurance Trainee Report

Discord: David#2177
Object of testing: #4892 
Tier of Testing: 1
Date: (25/05/2021)

Testing Process: Full Flight LIRN to LIEO

Departed LIRN with Trans ALT set in MCDU to FL80 on PERF page
After passing FL80 'QNH' flashes as expected on PFD
Completed a climb up to FL340
Inserted Trans ALT to 6000ft (for arrival)
Descended down to 4000ft. On passing 6000ft, STD flashes as expected on PFD

Negatives: N/A
Testing Results: Passed

Conclusions: Works as expected

**Comment by lukecologne** [comment]
> you need to add these code because it should not flashing when on ground

I don't think we need to explicitly prevent this from happening. This should never be the case if the TA/TL is entered correctly.

> also change these > < to <= >=

What difference would this make? The probably of the altitude being exactly equal to the TA/TL would be really small

@Kimbyeoungjang

**Comment by Kimbyeoungjang** [comment]
> > you need to add these code because it should not flashing when on ground
> 
> I don't think we need to explicitly prevent this from happening. This should never be the case if the TA/TL is entered correctly.
> 
> > also change these > < to <= >=
> 
> What difference would this make? The probably of the altitude being exactly equal to the TA/TL would be really small
> 
> @Kimbyeoungjang

Because IRL Pilot was confirmed to me it is inhibit on ground

And our current FMC default TA is 10000ft so if spawn on high elevation airport like tibet, it might flashing while on ground which is not corret.

And yes >= and > is quiet small different but this also confirmed by IRL pilot.
If we level off on TA or TL, it should flashing

**Comment by lukecologne** [comment]
> Because IRL Pilot was confirmed to me it is inhibit on ground
> 
> And our current FMC default TA is 10000ft so if spawn on high elevation airport like tibet, it might flashing while on ground which is not corret.
> 
> And yes >= and > is quiet small different but this also confirmed by IRL pilot.
> If we level off on TA or TL, it should flashing

Alright, I'll add the ground check. Can you give a link to the conversation on Discord? To the second point: I can assure you that you will not be realistically able to fly at exactly TA, there will pretty much always be a difference, even if you can't see it on the PFD. So the altitude was probably a bit over TA in that situation.

**Comment by Kimbyeoungjang** [comment]
> > Because IRL Pilot was confirmed to me it is inhibit on ground
> > And our current FMC default TA is 10000ft so if spawn on high elevation airport like tibet, it might flashing while on ground which is not corret.
> > And yes >= and > is quiet small different but this also confirmed by IRL pilot.
> > If we level off on TA or TL, it should flashing
> 
> Alright, I'll add the ground check. Can you give a link to the conversation on Discord? To the second point: I can assure you that you will not be realistically able to fly at exactly TA, there will pretty much always be a difference, even if you can't see it on the PFD. So the altitude was probably a bit over TA in that situation.

Screenshot_20210526-073233_Samsung Internet

Verrryyyyyyy old comment XD

The second point is very edge case because most ATC is not fly to TA or TL but look at that upper cooment from fankkop.

IRL seems flasing while on TA or TL

**Comment by lukecologne** [comment]
No offence but that source does not seem to be really that reliable. I don't think that user is veryfied on discord. To the second point: in the FCOM it states below/above TA, not below/above or equal to.

**Comment by Kimbyeoungjang** [comment]
> No offence but that source does not seem to be really that reliable. I don't think that user is veryfied on discord. To the second point: in the FCOM it states below/above TA, not below/above or equal to.

image
IRL confirm flashing on equal.

and I'll ask pilot about ground thing :D

**Comment by tracernz** [comment]
> > No offence but that source does not seem to be really that reliable. I don't think that user is veryfied on discord. To the second point: in the FCOM it states below/above TA, not below/above or equal to.
> 
> image
> IRL confirm flashing on equal.
> 
> and I'll ask pilot about ground thing :D

You cannot confirm strictly equal without hooking a real FMGC up to a test harness, because you can never keep a really plane at an exact altitude. We can keep it as Luke has done it.

**Comment by St54Kevin** [comment]
Does this need any more testing?

**Comment by tracernz** [comment]
> 
> 
> Does this need any more testing?

Nope. Ready for merge.

**Comment by FlyingCarpet7** [comment]
Just made a flight and during descent when changing from STD to QNH before reaching Trans Alt, the QNH starts flashing incorrectly all the way until passing Trans Alt. FMGS was in descend phase, and then also approach phase.
Short videoclip:
https://youtu.be/PXjv1PFzyGA

**Comment by lukecologne** [comment]
Interesting. Are you able replicate that? If yes, can you look up the FMGC flight phase simvar, transition alt and transition level (simvar is called something with APPR_TRANS_ALT I think) in the devmode?

**Comment by FlyingCarpet7** [comment]
I can try the same flight tomorrow. So I can find the simvars from the dev mode menu inside the sim then?

**Comment by lukecologne** [comment]
Yes. In the devmode bar, click on windows, model behaviors, then LocalVariables, and there you can type in the searchbar the vars I mentioned.

**Comment by FlyingCarpet7** [comment]
Same thing again.
Captured some simvars after I passed the trans alt and it stopped flashing:

Microsoft Flight Simulator Screenshot 2021 05 28 - 14 54 35 64 (2)
Microsoft Flight Simulator Screenshot 2021 05 28 - 14 54 24 52 (2)

**Comment by Kimbyeoungjang** [comment]
> Same thing again.
> Captured some simvars after I passed the trans alt and it stopped flashing:
> 
> Microsoft Flight Simulator Screenshot 2021 05 28 - 14 54 35 64 (2)
> Microsoft Flight Simulator Screenshot 2021 05 28 - 14 54 24 52 (2)

because code is wrong.
i'll fix that
